servicenow/stackstate_checks/servicenow/servicenow.py

```diff
@@ -167,73 +199,181 @@ def _process_components(self, components, instance_config):
             child_sys_id = relation['child']['value']
             type_sys_id = relation['type']['value']
 
             relation_type = relation_types[type_sys_id]
             data = self.filter_empty_metadata(relation)
-            data.update({"tags": instance_config.instance_tags})
+            data.update({"tags": instance_info.instance_tags})
 
             self.relation(parent_sys_id, child_sys_id, relation_type, data)
 
+    def _collect_change_requests(self, instance_info, latest_sys_updated_on=None):
+        # TODO: add limit for max number of events that we'll process
+        auth = (instance_info.user, instance_info.password)
+        url = instance_info.url + '/api/now/table/change_request'
+        if latest_sys_updated_on:
+            reformatted_date = ','.join("'{}'".format(i) for i in str(latest_sys_updated_on).split(' '))
+            quoted_date = urllib.parse.quote(reformatted_date)
+            params = '?sysparm_query=sys_updated_on>javascript%3Ags.dateGenerate({})'.format(quoted_date)
+            params += '&sysparm_display_value=true'
+            url += params
+        self.log.debug('url for getting CRs: %s', url)
+        return self._get_json(url, instance_info.timeout, auth)
+
+    def _sanitize_response(self, cr_list):
+        sanitized_crs = []
+        for cr in cr_list:
+            sanitized_crs.append(self.filter_empty_metadata(cr))
+        return sanitized_crs
+
+    def _process_change_requests(self, instance_info):
+        state = self.persistent_state.get_state(self.persistent_instance)
+        latest_sys_updated_on = datetime.datetime.strptime(state.get(instance_info.url), TIME_FORMAT)
+        response = self._collect_change_requests(instance_info, latest_sys_updated_on)
+        sanitized_result = self._sanitize_response(response['result'])
+        crs_persisted_state = state[self.cr_persistence_key]
+        for cr in sanitized_result:
+            try:
+                change_request = ChangeRequest(cr, strict=False)
+                change_request.validate()
+            except DataError as e:
+                self.log.info('%s - DataError: %s. This CR is skipped.', cr.get('number'), e)
+                continue
+            if change_request.cmdb_ci:
+                if change_request.sys_updated_on > latest_sys_updated_on:
+                    latest_sys_updated_on = change_request.sys_updated_on
+                    state[instance_info.url] = str(latest_sys_updated_on)
+                old_state = crs_persisted_state.get(change_request.number)
+                if old_state is None or old_state != change_request.state:
+                    self._create_event_from_change_request(change_request, instance_info)
+                    crs_persisted_state[change_request.number] = change_request.state
+        self.persistent_state.set_state(self.persistent_instance, state)
+
+    def _load_state(self, instance_info):
+        """
+        Persistent State structure:
+        {
```

Comment:
change the structure to 
{
      url: {
                  lastProcessedCRtimestamp: timestamp,
                  changeRequest: {
                                                      change_request_number: state
                                               }
             }

Reply:
done